dataflux_pytorch/multipart_upload/multipart.py

```diff
@@ -0,0 +1,313 @@
+ You may obtain a copy of the License at
+
+      https://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ """
+import concurrent.futures
+
+from google.api_core import exceptions
+from google.cloud.storage import Client
+from google.cloud.storage import Blob
+from google.cloud.storage.blob import _get_host_name
+from google.cloud.storage.blob import _quote
+from google.cloud.storage.constants import _DEFAULT_TIMEOUT
+from google.cloud.storage._helpers import _api_core_retry_to_resumable_media_retry
+from google.cloud.storage.retry import DEFAULT_RETRY
+from google.cloud.storage.transfer_manager import (
+    _api_core_retry_to_resumable_media_retry, )
+from google.cloud.storage.transfer_manager import _headers_from_metadata
+from google.cloud.storage.transfer_manager import _get_pool_class_and_requirements
+from google.resumable_media import _helpers
+
+import google_crc32c
+
+from google.resumable_media.requests.upload import XMLMPUContainer
+from google.resumable_media.requests.upload import XMLMPUPart
+from google.resumable_media.common import DataCorruption
+
+TM_DEFAULT_CHUNK_SIZE = 32 * 1024 * 1024
+DEFAULT_MAX_WORKERS = 8
+
+# Constants to be passed in as `worker_type`.
+PROCESS = "process"
+THREAD = "thread"
+
+
+def upload_chunks_concurrently_from_bytesio(
+    bytesio,
+    blob,
+    content_type=None,
+    chunk_size=TM_DEFAULT_CHUNK_SIZE,
+    deadline=None,
+    max_workers=DEFAULT_MAX_WORKERS,
+    *,
+    checksum="crc32c",
+    timeout=_DEFAULT_TIMEOUT,
+    retry=DEFAULT_RETRY,
+):
+    """Upload a single BytesIO object in chunks, concurrently.
+
+    This function uses the XML MPU API to initialize an upload and upload a
+    file in chunks, concurrently with a worker pool.
+
+    The XML MPU API is significantly different from other uploads; please review
+    the documentation at `https://cloud.google.com/storage/docs/multipart-uploads`
+    before using this feature.
+
+    The library will attempt to cancel uploads that fail due to an exception.
+    If the upload fails in a way that precludes cancellation, such as a
+    hardware failure, process termination, or power outage, then the incomplete
+    upload may persist indefinitely. To mitigate this, set the
+    `AbortIncompleteMultipartUpload` with a nonzero `Age` in bucket lifecycle
+    rules, or refer to the XML API documentation linked above to learn more
+    about how to list and delete individual downloads.
+
+    Using this feature with multiple threads is unlikely to improve upload
+    performance under normal circumstances due to Python interpreter threading
+    behavior. The default is therefore to use processes instead of threads.
```

Comment:
I wanted to get clarification here, the docs here suggest the default is to use processes, however we're using threads which makes sense for a network bound task. What is the default referring to in this paragraph?